pkg/engine/context/imageutils.go

```diff
@@ -35,10 +35,13 @@ type ImageInfo struct {
 
 func (i *ImageInfo) String() string {
 	image := i.Registry + "/" + i.Path + ":" + i.Tag
-	if i.Digest != "" {
+	// image that needs only digest and not the tag
+	imageWithoutTag := i.Registry + "/" + i.Path
+	if i.Digest != "" && i.Tag == "" {
+		image = imageWithoutTag + "@" + i.Digest
+	} else if i.Digest != "" && i.Tag != "" {
```

Comment:
This still seems incorrect as this will result in something like

`registry.io/path:@digest`

Notice the extra colon which is invalid and present due to the result of L37.

We should just use the above conditional where if the digest is non empty, we create the image as per L41 regardless of the tag value.

Reply:
Thanks a lot for the review and apologies for the delay in getting back. 
Understood, so this means that if we have a non-empty digest, we would not like to consider its tag value right? 
I tried the suggested code change but I guess it failed for the test case where we had both tag and digest non-empty and mentioned, should we remove this taste case considering the priority and security we get from using digest value when available? 

```
--- FAIL: Test_ImageInfo_String (0.00s)
    imageutils_test.go:133: 
                Error Trace:    imageutils_test.go:133
                                                        imageutils_test.go:62
                Error:          Not equal: 
                                expected: "registry.test.io/test/myapp:v1.2-21.g5523e95@sha256:31aaf12480bd08c54e7990c6b0e43d775a7a84603d2921a6de4abbc317b2fd10"
                                actual  : "registry.test.io/test/myapp@sha256:31aaf12480bd08c54e7990c6b0e43d775a7a84603d2921a6de4abbc317b2fd10"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -registry.test.io/test/myapp:v1.2-21.g5523e95@sha256:31aaf12480bd08c54e7990c6b0e43d775a7a84603d2921a6de4abbc317b2fd10
                                +registry.test.io/test/myapp@sha256:31aaf12480bd08c54e7990c6b0e43d775a7a84603d2921a6de4abbc317b2fd10
                Test:           Test_ImageInfo_String
           
 ```

Reply:
Have gone with that suggestion,  keeping digest over tag whenever available, do let me know if it works, thanks a lot for your review and feedback! :smile: